### PR TITLE
MULE-9994: Improve overall performance of JUnit tests

### PR DESCRIPTION
**Pull request by elrodro83**
* Scratch the surface

### CODE REVIEW

File: core-tests/src/test/java/org/mule/runtime/core/context/DefaultMuleContextTestCase.java [hunk 1]
```diff
@@ -108,6 +109,8 @@ public void callSystemExceptionHandlerWhenExceptionIsMessagingException() throws
         context.setExceptionListener(mockSystemExceptionHandler);
         context.handleException(mockMessagingException);
         verify(mockSystemExceptionHandler, VerificationModeFactory.times(1)).handleException(mockMessagingException,null);
+
+        context.dispose();
```
Comment (marianogonzalez) [inline core-tests/src/test/java/org/mule/runtime/core/context/DefaultMuleContextTestCase.java]: shouldn't this context be a class field and have an after() which cleans it up if not null?

File: core-tests/src/test/java/org/mule/runtime/core/registry/RegisteredObjectLifecycleTestCase.java [hunk 2]
```diff
@@ -44,18 +46,18 @@ public void testLifecycleForMuleContext() throws Exception
     {
         muleContext.getRegistry().registerObject("dummy", bean);
         assertTrue(initLatch.await(TIMEOUT, TimeUnit.MILLISECONDS));
-        assertFalse(startLatch.await(TIMEOUT, TimeUnit.MILLISECONDS));
-        assertFalse(stopLatch.await(TIMEOUT, TimeUnit.MILLISECONDS));
-        assertFalse(disposeLatch.await(TIMEOUT, TimeUnit.MILLISECONDS));
+        assertThat(startLatch.getCount(), is(1L));
```
Comment (marianogonzalez) [inline core-tests/src/test/java/org/mule/runtime/core/registry/RegisteredObjectLifecycleTestCase.java]: not the same... applies to the entire class

File: core-tests/src/test/java/org/mule/runtime/core/util/queue/QueueManagerLifecycleOrderTestCase.java [hunk 3]
```diff
@@ -44,6 +45,8 @@ public void testStartupOrder() throws Exception
         assertSame(fc, startStopOrder.get(1));
         assertSame(fc, startStopOrder.get(2));
         assertSame(rtqm, startStopOrder.get(3));
+
+        muleContext.dispose();
```
Comment (marianogonzalez) [inline core-tests/src/test/java/org/mule/runtime/core/util/queue/QueueManagerLifecycleOrderTestCase.java]: this should happen on an after()

File: core/src/main/java/org/mule/runtime/core/config/pool/ThreadPoolFactory.java [hunk 4]
```diff
@@ -34,17 +32,7 @@
...
-           it's hidden in earlier versions.
-
-           The javax.imageio.spi.ServiceRegistry, while not belonging here at first look, is
-           perfectly functional. Underneath, it wraps the sun.misc.Service class, which does the
-           lookup. The latter has been available since Java 1.3 and standardized internally at Sun.
-           Also, Sun, Bea JRockit and IBM JDK all have it readily available for use, so it's safe to
-           rely on.
-        */
-        final Iterator<ThreadPoolFactory> servicesIterator = ServiceRegistry.lookupProviders(ThreadPoolFactory.class);
+        final Iterator<ThreadPoolFactory> servicesIterator = load(ThreadPoolFactory.class).iterator();
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/config/pool/ThreadPoolFactory.java]: Use SpiServiceProvider instead

File: modules/spring-config/src/main/java/org/mule/runtime/config/spring/MuleDocumentLoader.java [hunk 5]
```diff
@@ -56,7 +60,26 @@ public void handle(short operation, String key, Object data, Node src, Node dst)
...
+    private final DocumentLoader defaultLoader = new DefaultDocumentLoader()
+    {
+        @Override
+        protected DocumentBuilderFactory createDocumentBuilderFactory(int validationMode, boolean namespaceAware)
+                throws ParserConfigurationException
+        {
+            DocumentBuilderFactory factory = super.createDocumentBuilderFactory(validationMode, namespaceAware);
+            try
+            {
+                factory.setFeature(DEFER_NODE_EXPANSION_FEATURE_KEY, false);
```
Comment (marianogonzalez) [inline modules/spring-config/src/main/java/org/mule/runtime/config/spring/MuleDocumentLoader.java]: we should run a profile on this before introducing it. For example, take  the most heavy metal templates and try it with thouse

Comment (elrodro83) [inline modules/spring-config/src/main/java/org/mule/runtime/config/spring/MuleDocumentLoader.java]: i can remove that change for now and analyze further if it's worth it.

File: core-tests/src/test/java/org/mule/runtime/core/context/DefaultMuleContextTestCase.java [hunk 6]
```diff
@@ -58,31 +57,42 @@
...
     @Before
     public void before()
     {
         muleContextFactory = new DefaultMuleContextFactory();
     }
 
+    @After
+    public void after() throws MuleException
+    {
+        if (context.isStarted())
```
Comment (marianogonzalez) [inline core-tests/src/test/java/org/mule/runtime/core/context/DefaultMuleContextTestCase.java]: shouldn't you also test the context to not be null?

File: core-tests/src/test/java/org/mule/runtime/core/registry/RegisteredObjectLifecycleTestCase.java [hunk 7]
```diff
@@ -46,17 +46,20 @@ public void testLifecycleForMuleContext() throws Exception
     {
         muleContext.getRegistry().registerObject("dummy", bean);
         assertTrue(initLatch.await(TIMEOUT, TimeUnit.MILLISECONDS));
+        Thread.sleep(TIMEOUT);
```
Comment (marianogonzalez) [inline core-tests/src/test/java/org/mule/runtime/core/registry/RegisteredObjectLifecycleTestCase.java]: replace sleeps with polling prober

Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/registry/RegisteredObjectLifecycleTestCase.java]: Actually, this is testing the state of the conditions after a certain time. With a prober, the first time it runs they would pass the assertion and we wouldn't be testing what happens after a certain time
